File: src/goryu/ofproto/ofaction.go
// -*- coding: utf-8 -*-

// Copyright (C) 2017 Nippon Telegraph and Telephone Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package ofproto

import (
	"fmt"
	"strings"
)

//
// Action interface
//
type Action interface {
}

//
// Action (Base)
//
type BaseAction struct {
	Name     string
	StrValue string
	IntValue int64
}

func (a *BaseAction) String() string {
	if a.IntValue == 0 {
		return fmt.Sprintf("%s:\"%s\"", a.Name, strings.Trim(a.StrValue, " "))
	} else {
		return fmt.Sprintf("%s:0x%x", a.Name, a.IntValue)
	}
}

//
// Action (GOTO_TABLE)
//
type GotoTableAction struct {
	TableNo uint8
}

func (a *GotoTableAction) String() string {
	return fmt.Sprintf("GOTO_TABLE:%02d(%s)", a.TableNo, StrTable(a.TableNo))
}

//
// Action (Group)
//
type GroupAction struct {
	GroupId uint32
}

func (a *GroupAction) String() string {
	return fmt.Sprintf("GROUP:%08x(%s)", a.GroupId, ConvGroupId(a.GroupId))
}

type Actions []map[string]interface{}

func NewAction0(t string) map[string]interface{} {
	return map[string]interface{}{
		"type": t,
	}
}

func NewAction(t string, key string, value interface{}) map[string]interface{} {
	return map[string]interface{}{
		"type": t,
		key:    value,
	}
}

func (a *Actions) Append(m map[string]interface{}) *Actions {
	*a = append(*a, m)
	return a
}

func (a *Actions) Output(port uint32) *Actions {
	return a.Append(NewAction("OUTPUT", "port", port))
}

func (a *Actions) CopyTTLOut() *Actions {
	return a.Append(NewAction0("COPY_TTL_OUT"))
}

func (a *Actions) CopyTTLIn() *Actions {
	return a.Append(NewAction0("COPY_TTL_IN"))
}

func (a *Actions) SetMPLSTTL(mpls_ttl uint8) *Actions {
	return a.Append(NewAction("SET_MPLS_TTL", "mpls_ttl", mpls_ttl))
}

func (a *Actions) DecMPLSTTL() *Actions {
	return a.Append(NewAction0("DEC_MPLS_TTL"))
}

func (a *Actions) PushVLAN(ethertype uint16) *Actions {
	return a.Append(NewAction("PUSH_VLAN", "ethertype", ethertype))
}

func (a *Actions) PopVLAN() *Actions {
	return a.Append(NewAction0("POP_VLAN"))
}

func (a *Actions) PushMPLS(ethertype uint16) *Actions {
	return a.Append(NewAction("PUSH_MPLS", "ethertype", ethertype))
}

func (a *Actions) PopMPLS(ethertype uint16) *Actions {
	return a.Append(NewAction("POP_MPLS", "ethertype", ethertype))
}

func (a *Actions) SetQueue(queue_id uint32) *Actions {
	return a.Append(NewAction("SET_QUEUE", "queue_id", queue_id))
}

func (a *Actions) Group(group_id uint32) *Actions {
	return a.Append(NewAction("GROUP", "group_id", group_id))
}

func (a *Actions) SetNWTTL(nw_ttl uint8) *Actions {
	return a.Append(NewAction("SET_NW_TTL", "nw_ttl", nw_ttl))
}

func (a *Actions) DecNWTTL() *Actions {
	return a.Append(NewAction0("DEC_NW_TTL"))
}

func (a *Actions) SetField(field string, value interface{}) *Actions {
	m := map[string]interface{}{
		"type":  "SET_FIELD",
		"field": field,
		"value": value,
	}
	return a.Append(m)
}

func (a *Actions) PushPBB(ethertype uint16) *Actions {
	return a.Append(NewAction("PUSH_PBB", "ethertype", ethertype))
}

func (a *Actions) PopPBB() *Actions {
	return a.Append(NewAction0("POP_PBB"))
}

func (a *Actions) WriteActions(actions Actions) *Actions {
	return a.Append(NewAction("WRITE_ACTIONS", "actions", actions))
}

func (a *Actions) ClearActions() *Actions {
	return a.Append(NewAction0("CLEAR_ACTIONS"))
}

func (a *Actions) GotoTable(table_id uint8) *Actions {
	return a.Append(NewAction("GOTO_TABLE", "table_id", table_id))
}

func (a *Actions) WriteMetadata(metadata uint64, metadata_mask uint64) *Actions {
	m := map[string]interface{}{
		"type":          "WRITE_METADATA",
		"metadata":      metadata,
		"metadata_mask": metadata_mask,
	}
	return a.Append(m)
}

func (a *Actions) Meter(meter_id uint32) *Actions {
	return a.Append(NewAction("METER", "meter_id", meter_id))
}

func (a *Actions) SetVRF(vrf uint8, useMetadata bool) *Actions {
	if useMetadata {
		return a.WriteMetadata(uint64(vrf)<<VRF_METADATA_SHIFT, VRF_METADATA_MASK)
	} else {
		return a.SetField("vrf", vrf)
	}
}

func (a *Actions) SetMPLSType(mpls_type uint8, useMetadata bool) *Actions {
	if useMetadata {
		return a.WriteMetadata(uint64(mpls_type)<<MPLSTYPE_METADATA_SHIFT, MPLSTYPE_METADATA_MASK)
	} else {
		return a.SetField("mpls_type", mpls_type)
	}
}

File: src/goryu/ofproto/ofaction.py
# -*- coding: utf-8 -*-

# Copyright (C) 2017 Nippon Telegraph and Telephone Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
FIBC Ryu message helper functions
"""

from goryu.ofproto import ofproto

def _action(action_type, **kwargs):
    dic = dict(type=action_type)
    dic.update(**kwargs)
    return dic

def output(port):
    """
    OUTPUT Action
    """
    return _action("OUTPUT", port=port)

def copy_ttl_out():
    """
    COPY_TTL_OUT Action
    """
    return _action("COPY_TTL_OUT")

def copy_ttl_in():
    """
    COPY_TTL_IN Action
    """
    return _action("COPY_TTL_IN")

def set_mpls_ttl(ttl):
    """
    SET_MPLS_TTL Action
    """
    return _action("SET_MPLS_TTL", mpls_ttl=ttl)

def dec_mpls_ttl():
    """
    DEC_MPLS_TTL Action
    """
    return _action("DEC_MPLS_TTL")

def push_vlan(ethertype):
    """
    PUSH_VLAN
    """
    return _action("PUSH_VLAN", ethertype=ethertype)

def pop_vlan():
    """
    POP_VLAN
    """
    return _action("POP_VLAN")

def push_mpls(ethertype):
    """
    PUSH_MPLS
    """
    return _action("PUSH_MPLS", ethertype=ethertype)

def pop_mpls(ethertype):
    """
    POP_MPLS
    """
    return _action("POP_MPLS", ethertype=ethertype)

def set_queue(queue_id):
    """
    SET_QUEUE
    """
    return _action("SET_QUEUE", queue_id=queue_id)

def group(group_id):
    """
    GROUP Action
    """
    return _action("GROUP", group_id=group_id)

def set_nw_ttl(nw_ttl):
    """
    SET_NW_TTL Action
    """
    return _action("SET_NW_TTL", nw_ttl=nw_ttl)

def dec_nw_ttl():
    """
    DEC_NW_TTL Action
    """
    return _action("DEC_NW_TTL")

def set_field(field, value):
    """
    SET_FIELD Action
    """
    return _action("SET_FIELD", field=field, value=value)

def push_pbb(ethertype):
    """
    PUSH_PBB Action
    """
    return _action("PUSH_PBB", ethertype=ethertype)

def pop_pbb():
    """
    POP_PBB Action
    """
    return _action("POP_PBB")

def write_actions(actions):
    """
    WRITE_ACTIONS Action
    """
    return _action("WRITE_ACTIONS", actions=actions)

def clear_actions():
    """
    CLEAR_ACTIONS Action
    """
    return _action("CLEAR_ACTIONS")

def goto_table(table_id):
    """
    GOTO_TABLE Action
    """
    return _action("GOTO_TABLE", table_id=table_id)

def write_metadata(metadata, metadata_mask):
    """
    WRITE_METADATA Action
    """
    return _action("WRITE_METADATA", metadata=metadata, metadata_mask=metadata_mask)

def meter(meter_id):
    """
    METER Action
    """
    return _action("METER", meter_id=meter_id)

def set_vrf(vrf, use_metadata):
    """
    SET_FIELD(vrf) Action
    """
    if use_metadata:
        return write_metadata(
            vrf << ofproto.VRF_METADATA_SHIFT, ofproto.VRF_METADATA_MASK)

    return set_field("vrf", vrf)

def set_mpls_type(mpls_type, use_metadata):
    """
    SET_FIELD(mpls_type) Action
    """
    if use_metadata:
        return write_metadata(
            mpls_type << ofproto.MPLSTYPE_METADATA_SHIFT, ofproto.MPLSTYPE_METADATA_MASK)

    return set_field("mpls_type", mpls_type)
